File: Linear/Static/Array/06_Sorting/Merge Sort/MergeSort.c
#include <stdio.h>
#include <unistd.h>

// Merges two subarrays of arr[]
// First subarray is arr[l...m]
// Second subarray is arr[m+1...r]
void merge(int arr[], int l, int m, int r) {
    int n1 = m - l + 1;
    int n2 = r - m;

    // Create temp arrays
    int L[n1], R[n2];

    // Copy data to temp arrays L[] and R[]
    for (int i = 0; i < n1; i++) {
        L[i] = arr[l + i];
    }
    for (int i = 0; i < n2; i++) {
        R[i] = arr[m + 1 + i];
    }
    
    // Merge the temp arrays back into arr[left...right]
    int i = 0, j = 0, k = l;

    // merge elements in sorted order
    while (i < n1 && j < n2) {
        if (L[i] <= R[j]) {
            arr[k++] = L[i++];
        } else {
            arr[k++] = R[j++];
        }
        
    }
    
    /* copy the remaining elements
    of L[], if there are any */
    while (i < n1) {
        arr[k++] = L[i++];
    }
    
    /* copy the remaining elements
    of R[], if there are any */
    while (j < n2) {
        arr[k++] = R[j++];
    }
}

void mergeSort(int arr[], int l, int r) {
    if (l < r) {
        int m = (l + r) / 2;
        mergeSort(arr, l, m);
        mergeSort(arr, m + 1, r);
        merge(arr, l, m, r);
    }
    
}

// Function to print an array
void printArray(int arr[], int n) {
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }    
}

int main() {
    int arr[] = { 38, 27, 43, 10 };
    int n = sizeof(arr) / sizeof(arr[0]);
    printf("Unsorted array: \n");
    printArray(arr, n);

    for (int i = 3; i > 0; i--) {
        printf("\nSorting in %d", i);
        sleep(1);
    }

    mergeSort(arr, 0, n - 1);
    printf("\nSorted array: \n");
    printArray(arr, n);

    return 0;
}

File: Linear/Static/Array/06_Sorting/Merge Sort/MergeSort.py
import time

# Merges two subarrays of arr[]
# First subarray is arr[l...m]
# Second subarray is arr[m+1...r]
def merge(arr, l, m, r):
    n1 = m - l + 1
    n2 = r - m

    # Create temp arrays
    L = []
    R = []
    ARR = []

    # Copy data to temp arrays L[] and R[]
    for i in range(n1):
        L.append(arr[l + i])
    for i in range(n2):
        R.append(arr[m + 1 + i])
    
    # Merge the temp arrays back into arr[left...right]
    i = j = 0

    # merge elements in sorted order
    while i < n1 and j < n2:
        if L[i] <= R[j]:
            ARR.append(L[i])
            i += 1
        else:
            ARR.append(R[j])
            j += 1
        
    """copy remaining elements
    from L[], if there are any"""
    while i < n1:
        ARR.append(L[i])
        i += 1

    """copy remaining elements
    from R[], if there are any"""
    while j < n2:
        ARR.append(R[j])
        j += 1
    
    for i in range(len(ARR)):
        arr[l + i] = ARR[i]
    
    return arr

def mergeSort(arr, l, r):
    if l < r:
        m = (l + r) // 2
        arr = mergeSort(arr, l, m)
        arr = mergeSort(arr, m + 1, r)
        arr = merge(arr, l, m, r)
    
    return arr

# Function to print an array
def printArray(arr):
    for i in range(len(arr)):
        print(arr[i], end=" ")

arr = [ 38, 27, 43, 10 ]
print("Unsorted array:")
printArray(arr)

for i in range(3, 0, -1):
    print("\nSorting in", i)
    time.sleep(1)

arr = mergeSort(arr, 0, len(arr) - 1)
print("Sorted array:")
printArray(arr)